### PR TITLE
Add Mixtral models

### PR DESCRIPTION
**Pull request by artek0chumak**
Add [Mixtral models](https://huggingface.co/docs/transformers/model_doc/mixtral) as runnable models.

TODO
- [x] Wait for support of the new version of transformers  (>4.35)
  - [x] Add *normal* support for [Cache](https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py)
- [x] Check compatibility of larger models ([TinyMixtral](https://huggingface.co/SanjiWatsuki/TinyMixtral-32x248M), [Mixtral-8x7B](https://huggingface.co/mistralai/Mixtral-8x7B-v0.1)) on GPU
- [x] Refactor some code(see `# TODO`)
- [ ] Optimize code for improved speed (mb next PR?)

### CODE REVIEW

File: src/petals/server/from_pretrained.py [hunk 1]
```diff
@@ -51,7 +52,11 @@ def load_pretrained_block(
     torch_dtype = resolve_block_dtype(config, torch_dtype)
 
     with init_empty_weights():
-        block = config.block_class(config)
+        if config.block_class == WrappedMixtralBlock:
+            config = PreTrainedModel._autoset_attn_implementation(config)
```
Comment (artek0chumak) [inline src/petals/server/from_pretrained.py:56]: Called function: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L1316

File: src/petals/server/backend.py [hunk 2]
```diff
@@ -91,6 +91,8 @@ def get_inference_cache_descriptors(self, batch_size: int, max_length: int) -> S
         cache_tensors = []
         for device, num_heads in zip(self.module.devices, self.shard_num_heads):
             num_heads //= self.config.num_key_value_groups
+            if hasattr(self.config, "num_key_value_heads"):
+                num_heads = self.config.num_key_value_heads
```
Comment (mryab) [inline src/petals/server/backend.py:95]: Does this still work when the layer is sharded?

File: src/petals/models/mixtral/model.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    def word_embeddings_layernorm(self) -> nn.Module:  # For compatibility with RemoteGenerationMixin
+        return nn.Identity()
+
+    @property
+    def h(self) -> RemoteSequential:  # For compatibility with RemoteGenerationMixin
+        return self.layers
+
+    @property
+    def ln_f(self) -> nn.Module:  # For compatibility with RemoteGenerationMixin
+        return self.norm
```
Comment (mryab) [inline src/petals/models/mixtral/model.py:135]: Perhaps I couldn't find it, but where exactly is it used in RemoteGenerationMixin?

File: src/petals/models/mixtral/model.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            attentions=None,
+        )
+
+    @property
+    def word_embeddings(self) -> nn.Embedding:  # For compatibility with RemoteGenerationMixin
+        return self.embed_tokens
+
+    @property
+    def word_embeddings_layernorm(self) -> nn.Module:  # For compatibility with RemoteGenerationMixin
+        return nn.Identity()
```
Comment (mryab) [inline src/petals/models/mixtral/model.py:127]: Same here

File: src/petals/models/mixtral/model.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
+from typing import Optional
+
+import hivemind
```
Comment (mryab) [inline src/petals/models/mixtral/model.py:3]: Best to import DHT explicitly

File: src/petals/server/backend.py (hunk 2 above)
Comment (justheuristic) [reply src/petals/server/backend.py:95]: This works if layer is sharded, unless the user manually rewrites petals to specify an incompatible sharding config into TensorParallel. Without that, TensorParallel will use a default (automatic) config where
- only linear layers are sharded
- each shard computes the full attention kernel, relu, etc


If user shards the Mixtral model, it will be using default config and run  @artek0chumak  code correctly. Because each layer shard requires full KV cache and has full number of heads in config.


In other words, the current code will work in all cases except when user goes out of their way to modify TensorParallel config in a way that breaks petals servers, which would require changing run_server code.

File: src/petals/models/mixtral/block.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+)
+from transformers.models.mixtral.modeling_mixtral import MixtralDecoderLayer, MixtralModel
+
+
+class WrappedMixtralBlock(MixtralDecoderLayer):
+    def __init__(self, config: MixtralConfig, layer_idx: int):
+        self._attn_implementation = config._attn_implementation
+        self.sliding_window = config.sliding_window
+        self.layer_idx = layer_idx
+        super().__init__(config, layer_idx)
```
Comment (mryab) [inline src/petals/models/mixtral/block.py:18]:
```suggestion
        super().__init__(config, layer_idx)
        self._attn_implementation = config._attn_implementation
        self.sliding_window = config.sliding_window
        self.layer_idx = layer_idx
```

Comment (mryab) [reply src/petals/models/mixtral/block.py:18]: Init of the parent class is usually the first step, any reason why we can't make it that way?

File: src/petals/models/mixtral/model.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+from petals.client.remote_generation import RemoteGenerationMixin, RemotePastKeyValues
+from petals.client.remote_sequential import RemoteSequential
+from petals.models.mixtral.config import DistributedMixtralConfig
+from petals.utils.auto_config import DefaultRevisionMixin
+
+logger = get_logger(__name__)
+
+
+class DistributedMixtralModel(DefaultRevisionMixin, FromPretrainedMixin, PTuneMixin, MixtralModel):
+    """MixtalModel, but all transformer layers are hosted by the swarm"""
```
Comment (mryab) [inline src/petals/models/mixtral/model.py:27]:
```suggestion
    """MixtralModel, but all transformer layers are hosted by the swarm"""
```

File: src/petals/models/mixtral/block.py (hunk 6 above)
Comment (artek0chumak) [reply src/petals/models/mixtral/block.py]: No reason, I've changed the order to the common way.

**Review by mryab — APPROVED** [review]
